src/components/views/FootnotesSelector.tsx

```diff
@@ -0,0 +1,224 @@
+  notes: FootnoteWithIndex[]
+  inlineFootnote?: InlineFootnoteNode
+  onAdd: () => void
+  onInsert: (notes: FootnoteWithIndex[]) => void
+  onCancel: () => void
+  addNewLabel?: string
+}> = ({ notes, inlineFootnote, onAdd, onInsert, onCancel, addNewLabel }) => {
+  let selectedNotesMap
+
+  if (inlineFootnote) {
+    const rids = inlineFootnote.attrs.rids
+    const selectedNotes = notes.filter(({ node }) =>
+      rids.includes(node.attrs.id)
+    )
+    selectedNotesMap = new Map(
+      selectedNotes.map(({ node }) => [node.attrs.id, node])
+    )
+  }
+
+  const [selections, setSelections] = useState(
+    new Map<string, FootnoteNode>(selectedNotesMap)
+  )
+
+  const toggleSelection = (item: FootnoteNode) => {
+    const id = item.attrs.id
+    if (selections.has(id)) {
+      selections.delete(id)
+      setSelections(new Map([...selections]))
+    } else {
+      selections.set(id, item)
+      setSelections(new Map([...selections]))
+    }
+  }
+
+  const isSelected = (item: FootnoteNode) => {
+    return selections.has(item.attrs.id)
+  }
+
+  const handleClick = () => {
+    return onInsert(notes.filter(({ node }) => selections.has(node.attrs.id)))
+  }
+
+  return (
+    <Container>
+      <NotesContainer>
+        <FootnotesList
+          notes={notes}
+          inlineFootnote={inlineFootnote}
+          isSelected={isSelected}
+          onSelect={toggleSelection}
+        />
+      </NotesContainer>
+      <Actions>
+        <AddNewFootnote>
+          <IconTextButton onClick={onAdd}>
+            <AddNewIcon />
+            {addNewLabel || 'Add new'}
+          </IconTextButton>
+        </AddNewFootnote>
+        <ButtonGroup>
+          <SecondaryButton onClick={onCancel}>Cancel</SecondaryButton>
+          <PrimaryButton
+            onClick={handleClick}
+            disabled={selections.size === 0 && !inlineFootnote}
+          >
+            {inlineFootnote ? 'Update' : 'Insert'}
+          </PrimaryButton>
+        </ButtonGroup>
+      </Actions>
+    </Container>
+  )
+}
+
+const FootnotesList: React.FC<{
+  notes: FootnoteWithIndex[]
+  inlineFootnote?: InlineFootnoteNode
+  isSelected: (item: FootnoteNode) => boolean
+  onSelect: (item: FootnoteNode) => void
+}> = ({ notes, isSelected, onSelect, inlineFootnote }) => {
+  const selectedNotes: FootnoteWithIndex[] = []
+  const remainingNotes: FootnoteWithIndex[] = []
+
+  notes.forEach((note) => {
+    const isNoteSelected =
+      inlineFootnote && inlineFootnote.attrs.rids.includes(note.node.attrs.id)
+    if (isNoteSelected) {
+      selectedNotes.push(note)
+    } else {
+      remainingNotes.push(note)
+    }
+  })
+
+  return (
+    <NotesListContainer>
+      {selectedNotes.map((note) => footnoteItem(note, isSelected, onSelect))}
+      {selectedNotes.length > 0 && remainingNotes.length > 0 && <Separator />}
+      {remainingNotes.map((note) => footnoteItem(note, isSelected, onSelect))}
+    </NotesListContainer>
+  )
+}
+
+const footnoteItem = (
```

Comment:
since we move it here, maybe we can fix it to be a proper react component?  Not sure why we actually use it as a function and no jsx component.

Reply:
Updated.